core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -356,4 +374,233 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
+  def partitionPorts(
+      conf: SparkConf,
+      ports: List[Resource])
+    : (List[Resource], List[Resource], List[Long]) = {
+    val taskPortRanges = getRangeResourceWithRoleInfo(ports.asJava, "ports")
+    val portsToCheck = ManagedPorts.getPortValues(conf)
+    val nonZeroPorts = portsToCheck.filter(_ != 0)
+    // reserve non zero ports first
+    val nonZeroResources = reservePorts(taskPortRanges, nonZeroPorts)
+    // reserve actual port numbers for zero ports - not set by the user
+    val numOfZeroPorts = portsToCheck.count(_ == 0)
+    val randPorts = pickRandomPortsFromRanges(nonZeroResources._1, numOfZeroPorts)
+    val zeroResources = reservePorts(nonZeroResources._1, randPorts)
+    val (portResourcesLeft, portResourcesToBeUsed) =
+      createResources(nonZeroResources, zeroResources)
+    (portResourcesLeft, portResourcesToBeUsed, nonZeroPorts ++ randPorts)
+  }
+
+  private object ManagedPorts {
+    val portNames = List("spark.executor.port", "spark.blockManager.port")
+
+    def getPortValues(conf: SparkConf): List[Long] = {
+      portNames.map(conf.getLong(_, 0))
+    }
+  }
+
+  private def createResources(
+      nonZero: (List[PortRangeResourceInfo], List[PortRangeResourceInfo]),
+      zero: (List[PortRangeResourceInfo], List[PortRangeResourceInfo]))
+    : (List[Resource], List[Resource]) = {
+    val resources = {
+      if (nonZero._2.isEmpty) { // no user ports were defined
+        (zero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))},
+          zero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))})
+
+      } else if (zero._2.isEmpty) { // no random ports were defined
+        (nonZero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))},
+          nonZero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))})
+      }
+      else {  // we have user defined and random ports defined
+        val left = zero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))}
+        val used = nonZero._2.flatMap{port =>
+          createMesosPortResource(port.range, Some(port.role))} ++
+          zero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))}
+        (left, used)
+      }
+    }
+    resources
+  }
+
+  private case class PortRangeResourceInfo(role: String, range: List[(Long, Long)])
+
+  private def getRangeResourceWithRoleInfo(res: JList[Resource], name: String)
+    : List[PortRangeResourceInfo] = {
+    // A resource can have multiple values in the offer since it can either be from
+    // a specific role or wildcard.
+    // Extract role info and port range for every port resource in the offer
+    res.asScala.filter(_.getName == name)
+      .map{res => PortRangeResourceInfo(res.getRole, res.getRanges.getRangeList.asScala
+        .map(r => (r.getBegin, r.getEnd)).toList) }.toList
+  }
+
+  /** Helper method to get a pair of assigned and remaining ports along with role info */
```

Comment:
I used a javadoc comment as here: https://github.com/databricks/scala-style-guide#doc
but since is private i didnt want to create the full documentation for the method. Do you prefer the full doc even for private methods?